src/lib/operators/insert.cpp

```diff
@@ -107,136 +93,132 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
+        source_row_id.chunk_offset = 0;
+      } else {
+        source_row_id.chunk_offset += num_rows_current_iteration;
       }
-    }
 
-    for (auto i = start_index; i < start_index + current_num_rows_to_insert; i++) {
-      // we do not need to check whether other operators have locked the rows, we have just created them
-      // and they are not visible for other operators.
-      // the transaction IDs are set here and not during the resize, because
-      // tbb::concurrent_vector::grow_to_at_least(n, t)" does not work with atomics, since their copy constructor is
-      // deleted.
-      target_chunk->get_scoped_mvcc_data_lock()->tids[i] = context->transaction_id();
-      _inserted_rows.emplace_back(RowID{target_chunk_id, i});
+      target_chunk_offset += num_rows_current_iteration;
+      target_chunk_range_remaining_rows -= num_rows_current_iteration;
     }
-
-    input_offset += current_num_rows_to_insert;
-    start_index = 0u;
   }
 
   return nullptr;
 }
 
 void Insert::_on_commit_records(const CommitID cid) {
-  for (auto row_id : _inserted_rows) {
-    auto chunk = _target_table->get_chunk(row_id.chunk_id);
-
-    auto mvcc_data = chunk->get_scoped_mvcc_data_lock();
-    mvcc_data->begin_cids[row_id.chunk_offset] = cid;
-    mvcc_data->tids[row_id.chunk_offset] = 0u;
+  for (const auto& target_chunk_range : _target_chunk_ranges) {
+    const auto target_chunk = _target_table->get_chunk(target_chunk_range.chunk_id);
+    auto mvcc_data = target_chunk->get_scoped_mvcc_data_lock();
+
+    for (auto chunk_offset = target_chunk_range.begin_chunk_offset; chunk_offset < target_chunk_range.end_chunk_offset;
+         ++chunk_offset) {
+      mvcc_data->begin_cids[chunk_offset] = cid;
+      mvcc_data->tids[chunk_offset] = 0u;
+    }
   }
 }
 
 void Insert::_on_rollback_records() {
-  for (auto row_id : _inserted_rows) {
-    auto chunk = _target_table->get_chunk(row_id.chunk_id);
-    // We set the begin and end cids to 0 (effectively making it invisible for everyone) so that the ChunkCompression
-    // does not think that this row is still incomplete. We need to make sure that the end is written before the begin.
-    chunk->get_scoped_mvcc_data_lock()->end_cids[row_id.chunk_offset] = 0u;
-    std::atomic_thread_fence(std::memory_order_release);
-    chunk->get_scoped_mvcc_data_lock()->begin_cids[row_id.chunk_offset] = 0u;
-
-    chunk->get_scoped_mvcc_data_lock()->tids[row_id.chunk_offset] = 0u;
+  for (const auto& target_chunk_range : _target_chunk_ranges) {
+    const auto target_chunk = _target_table->get_chunk(target_chunk_range.chunk_id);
+    auto mvcc_data = target_chunk->get_scoped_mvcc_data_lock();
+
+    for (auto chunk_offset = target_chunk_range.begin_chunk_offset; chunk_offset < target_chunk_range.end_chunk_offset;
+         ++chunk_offset) {
+      // Set the begin and end cids to 0 (effectively making it invisible for everyone).
+      // Make sure that the end is written before the begin.
```

Comment:
Write end_cids first, issue a single fence, then write the others?

Reply:
Why the fencing? From around 5h of watching c++ memory model talks and very little security gained on the topic I'd say `mvcc_data->tids[chunk_offset] = 0u;` at the very end is the release operation that makes the inserted row visible. No fences needed....

Reply:
You added the fencing. My suggestion is to remove it from the loop for performance reasons.

Reply:
the fence was there before, I doubt it was I who added it originally.

Reply:
Ah, ok. As this PR rewrote basically everything, I used split view and didn't check the left side too much.

Reply:
If we wrote the begin CID before we write the end CID, we might end up in a situation where the row (with begin 0, end infty) is indisinguishable from a row that is visible to everyone. This should be added to the comment.

Reply:
Added this comment, what do you think?

```c++
    /**
     * !!! Crucial comment, PLEASE READ AND _UNDERSTAND_ before altering any of the following code !!!
     *
     * Set end_cids to 0 (effectively making the rows invisible for everyone) BEFORE setting the begin_cids to 0.
     *
     * Otherwise, another thread might observe a column with `begin_cid == 0` and `end_cid == MAX_COMMIT_ID` which would
     * make this other thread believe it can see a row that was actually rolled back and should never become visible.
     *
     * We need to set `begin_cid = 0` so that the ChunkCompressionTask can identify "completed" Chunks.
     */
```